proto-lens/src/Data/ProtoLens/Encoding.hs

```diff
@@ -53,20 +53,23 @@ parseMessage :: forall msg . Message msg => Parser () -> Parser msg
 parseMessage end = do
     (msg, unsetFields) <- loop def requiredFields
     if Map.null unsetFields
-        then return $ reverseRepeatedFields fields msg
+        then return $ over unknownFields reverse
+                    $ reverseRepeatedFields fields msg
         else fail $ "Missing required fields "
                         ++ show (map fieldDescriptorName
                                     $ Map.elems $ unsetFields)
   where
     fields = fieldsByTag descriptor
+    addUnknown :: TaggedValue -> msg -> msg
+    addUnknown !f = over unknownFields (\(!xs) -> f : xs)
     requiredFields = Map.filter isRequired fields
     loop :: msg -> Map.Map Tag (FieldDescriptor msg)
             -> Parser (msg, Map.Map Tag (FieldDescriptor msg))
     loop msg unsetFields = ((msg, unsetFields) <$ end)
                 <|> do
                     tv@(TaggedValue tag _) <- getTaggedValue
                     case Map.lookup tag fields of
-                        Nothing -> loop msg unsetFields
+                        Nothing -> (loop $! addUnknown tv msg) unsetFields
```

Comment:
why the strictness? to avoid big thunks if a big message is entirely unknown?

Reply:
Yeah, pretty much.  In particular, the `addUnknown` thunk might otherwise hold onto the old value of `msg`, which can be painful if there are a bunch of unknown fields in a row (e.g.: an unknown, unpacked repeated field).